libs/utils/src/http/endpoint.rs

```diff
@@ -63,9 +88,53 @@ async fn prometheus_metrics_handler(_req: Request<Body>) -> Result<Response<Body
     Ok(response)
 }
 
+pub fn add_request_id_middleware<B: hyper::body::HttpBody + Send + Sync + 'static>(
+) -> Middleware<B, ApiError> {
+    Middleware::pre(move |req| async move {
+        let request_id = match req.headers().get(&X_REQUEST_ID_HEADER) {
+            Some(request_id) => {
+                Cow::Borrowed(request_id.to_str().expect("extract request id value"))
+            }
+            None => {
+                let request_id = uuid::Uuid::new_v4();
+                Cow::Owned(request_id.to_string())
+            }
+        };
+
+        req.set_context(RequestId {
+            id: request_id.to_string(),
```

Comment:
Here we copy the value anyway, so Cow is no longer saves us an allocation. Sorry for going back and forth with it